File: examples/simple_dll.lua
--[[
   Copyright (c) 2015-present, Facebook, Inc.
   All rights reserved.
   This source code is licensed under the BSD-style license found in the
   LICENSE file in the root directory of this source tree. An additional grant
   of patent rights can be found in the PATENTS file in the same directory.
]]--

-- attacks the closest units, th simple_dll.lua [-t $hostname] [-p $port]
local DEBUG = 0 -- can take values 0, 1, 2 (from no output to most verbose)
local MICRO_MODE = true -- set to false for normal maps!
require 'torch'
torch.setdefaulttensortype('torch.FloatTensor')
require 'sys'
local lapp = require 'pl.lapp'
local args = lapp [[
Baselines for Starcraft
    -t,--hostname       (default "")    Give hostname / ip pointing to VM
    -h,--heuristic      (default "wc")  Check comments for list of heuristics
    -p,--port           (default 11111) Port for torchcraft. Do 0 to grab vm from cloud
]]

local skip_frames = 7
local port = args.port
local hostname = args.hostname or ""
print("hostname:", hostname)
print("port:", port)

require 'progressbar'
local progress = ProgressBar(-1)
progress.mutex = {lock = function() end, unlock = function() end} -- hack
progress:singleThreaded()

local tc = require 'torchcraft'
tc.DEBUG = DEBUG
local utils = require 'torchcraft.utils'

local function get_closest(position, unitsTable)
    local min_d = 1E30
    local closest_uid = nil
    for uid, ut in pairs(unitsTable) do
        local tmp_d = utils.distance(position, ut['position'])
        if tmp_d < min_d then
            min_d = tmp_d
            closest_uid = uid
        end
    end
    return closest_uid
end

local battles_won = 0
local battles_game = 0
local total_battles = 0

-- no ',' accepted in map names!
-- All paths must be relative to C:/StarCraft
local maps = {'Maps/BroodWar/micro/dragoons_zealots.scm',
              'Maps/BroodWar/micro/m5v5_c_far.scm'}

tc.mode.micro_battles = MICRO_MODE
tc.mode.replay = false
local nrestarts = -1

while total_battles < 40 do

    print("")
    print("CTRL-C to stop")
    print("")

    local frames_in_battle = 1
    local nloop = 1
    battles_won = 0
    battles_game = 0
    nrestarts = nrestarts + 1

    tc:init(hostname, port)
    local update = tc:connect(port)
    if DEBUG > 1 then
        print('Received init: ', update)
    end

    -- first message to BWAPI's side is setting up variables
    local setup = {
        tc.command(tc.set_speed, 0), tc.command(tc.set_gui, 1),
        tc.command(tc.set_cmd_optim, 1),
    }
    tc:send({table.concat(setup, ':')})

    local built_barracks = 0

    local tm = torch.Timer()
    while not tc.state.game_ended do
        progress:add('Loop', nloop, '%5d')
        progress:add('FPS', 1 / tm:time().real, '%5d')
        progress:add('WR', battles_won / (battles_game+1E-6), '%1.3f')
        progress:add('#Wins', battles_won, '%4d')
        progress:add('#Bttls', battles_game, '%4d')
        progress:add('Tot Bttls', total_battles, '%4d')
        progress:push()
        tm:reset()

        update = tc:receive()
        if DEBUG > 1 then
            print('Received update: ', update)
        end

        nloop = nloop + 1
        local actions = {}
        if tc.state.game_ended then
            break
        elseif tc.state.battle_just_ended then
            if DEBUG > 0 then
                print("BATTLE ENDED")
            end
            if tc.state.battle_won then -- we won (in micro battles)
                battles_won = battles_won + 1
            end
            battles_game = battles_game + 1
            total_battles = total_battles + 1
            frames_in_battle = 0
            if battles_game >= 10 then
                actions = {
                    tc.command(tc.set_map, maps[(nrestarts % #maps) + 1]),
                    tc.command(tc.restart),
                }
            end
        elseif tc.state.waiting_for_restart then
            -- a battle finished, waiting for the next one to start!
            if DEBUG > 0 then
                print("WAITING FOR RESTART")
            end
        else
            if tc.state.battle_frame_count % skip_frames == 0 then
                for uid, ut in pairs(tc.state.units_myself) do
                    if tc:isbuilding(ut.type) then -- tests production
                        if ut.type == tc.unittypes.Terran_Barracks then
                            table.insert(actions,
                            tc.command(tc.command_unit, uid, tc.cmd.Train,
                            tc.unittypes.Terran_Marine))
                        end
                    elseif tc:isworker(ut.type) then
                        if tc.state.resources_myself.ore >= 150
                            and tc.state.frame_from_bwapi - built_barracks > 240 then -- tests building
                            built_barracks = tc.state.frame_from_bwapi
                            local _, pos = next(tc:filter_type(
                            tc.state.units_myself,
                            {tc.unittypes.Terran_Command_Center}))
                            if pos ~= nil then pos = pos.position end
                            if pos ~= nil and not utils.is_in(ut.order,
                                tc.command2order[tc.unitcommandtypes.Build]) 
                                and not utils.is_in(ut.order,
                                tc.command2order[tc.unitcommandtypes.Right_Click_Position]) then
                                table.insert(actions,
                                tc.command(tc.command_unit, uid,
                                tc.cmd.Build, -1,
                                pos[1], pos[2] + 8, tc.unittypes.Terran_Barracks))
                            end
                        else -- tests gathering
                            if not utils.is_in(ut.order,
                                  tc.command2order[tc.unitcommandtypes.Gather])
                                  and not utils.is_in(ut.order,
                                  tc.command2order[tc.unitcommandtypes.Build])
                                  and not utils.is_in(ut.order,
                                  tc.command2order[tc.unitcommandtypes.Right_Click_Position]) then
                                -- avoid spamming the order is the unit is already 
                                -- following the right order or building!
                                local target = get_closest(ut.position,
                                    tc:filter_type(tc.state.units_neutral,
                                        {tc.unittypes.Resource_Mineral_Field,
                                         tc.unittypes.Resource_Mineral_Field_Type_2,
                                         tc.unittypes.Resource_Mineral_Field_Type_3}))
                                if target ~= nil then
                                    table.insert(actions,
                                    tc.command(tc.command_unit_protected, uid,
                                    tc.cmd.Right_Click_Unit, target))
                                end
                            end
                        end
                    else -- attacks closest
                        local target = get_closest(ut.position,
                                                   tc.state.units_enemy)
                        if target ~= nil then
                            table.insert(actions,
                            tc.command(tc.command_unit_protected, uid,
                            tc.cmd.Attack_Unit, target))
                        end
                    end
                end
                if frames_in_battle > 2*60*24 then -- quit after ~ 2 hours
                    actions = {tc.command(tc.quit)}
                    nrestarts = nrestarts + 1
                end
                progress:pop()
            end
        end

        if DEBUG > 1 then
            print("")
            print("Sending actions:")
            print(actions)
        end
        tc:send({table.concat(actions, ':')})
    end
    tc:close()
    sys.sleep(0.5)
    print()
    progress:reset()
    print("")
    collectgarbage()
    collectgarbage()
end
print("")

File: examples/cpp/simple_dll.cpp
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

#include <algorithm>
#include <cfloat>
#include <chrono>
#include <iostream>
#include <thread>

#include <gflags/gflags.h>
#include <torchcraft/client.h>
#include <torchcraft/constants.h>
#include <torchcraft/state.h>

namespace tc = torchcraft;

// CLI flags
DEFINE_string(hostname, "localhost", "Host running torchcraft server");
DEFINE_int32(port, 11111, "Port for torchcraft");
DEFINE_string(heuristic, "wc", "Heuristic");
DEFINE_int32(skip_frames, 7, "Take action every n frames");
DEFINE_bool(micro_mode, true, "Set to false for normal maps");

// Determine the closest unit to a given position
template <typename It>
It getClosest(int x, int y, It first, It last) {
  It closest = last;
  float mind = FLT_MAX;
  while (first != last) {
    float d = float(x - first->x) * (x - first->x) +
        float(y - first->y) * (y - first->y);
    if (d < mind) {
      closest = first;
      mind = d;
    }
    ++first;
  }
  return closest;
}

// Retrieve all units that are mineral fields
std::vector<tc::replayer::Unit> getMineralFields(tc::State* state) {
  auto neutralUnits = state->units[state->neutral_id];
  std::vector<tc::replayer::Unit> fields;
  std::copy_if(
      neutralUnits.begin(),
      neutralUnits.end(),
      std::back_inserter(fields),
      [](const tc::replayer::Unit& u) {
        return (
            u.type == tc::BW::UnitType::Resource_Mineral_Field ||
            u.type == tc::BW::UnitType::Resource_Mineral_Field_Type_2 ||
            u.type == tc::BW::UnitType::Resource_Mineral_Field_Type_3);
      });
  return fields;
}

// Check whether a unit's current orders include the given command
bool isExecutingCommand(
    const tc::replayer::Unit& unit,
    tc::BW::UnitCommandType command) {
  auto orders = tc::BW::commandToOrders(command);
  auto res = std::find_first_of(
      unit.orders.begin(),
      unit.orders.end(),
      orders.begin(),
      orders.end(),
      [](const tc::replayer::Order& o1, tc::BW::Order o2) {
        return o1.type == o2;
      });
  return res != unit.orders.end();
}

// Establish connection and perform initial handshake
std::unique_ptr<tc::Client> makeClient() {
  // Establish connection
  auto client = std::unique_ptr<tc::Client>(new tc::Client());
  if (!client->connect(FLAGS_hostname, FLAGS_port, 30000)) {
    throw std::runtime_error(
        std::string("Error establishing connection: ") + client->error());
  }

  // Perform handshake
  tc::Client::Options opts;
  opts.micro_battles = FLAGS_micro_mode;
  std::vector<std::string> upd;
  if (!client->init(upd, opts)) {
    throw std::runtime_error(
        std::string("Error initializing connection: ") + client->error());
  }
  if (client->state()->replay) {
    throw std::runtime_error("Expected non-replay map");
  }

  return client;
}

void playGame(std::shared_ptr<tc::Client> cl, int* totalBattles) {
  // First message: set up some variables
  std::vector<tc::Client::Command> comms;
  comms.emplace_back(tc::BW::Command::SetSpeed, 0);
  comms.emplace_back(tc::BW::Command::SetGui, 1);
  comms.emplace_back(tc::BW::Command::SetCmdOptim, 1);
  if (!cl->send(comms)) {
    throw std::runtime_error(std::string("Sent failure: ") + cl->error());
  }

  int builtBarracks = 0;
  int battlesWon = 0;
  int battlesGame = 0;
  int nloop = 0;

  auto last = std::chrono::system_clock::now();
  while (!cl->state()->game_ended) {
    // Display progress
    printf("Loop = %5d | ", nloop);
    printf(
        "FPS = %5lld | ",
        CLOCKS_PER_SEC / (std::chrono::system_clock::now() - last).count());
    printf("WR = %1.3f | ", battlesWon / (battlesGame + 1e-6));
    printf("#Wins = %4d | ", battlesWon);
    printf("#Bttls = %4d | ", battlesGame);
    printf("Tot Bttls = %4d |    ", *totalBattles);
    printf("\r");
    std::cout << std::flush;
    last = std::chrono::system_clock::now();

    std::vector<std::string> updates;
    if (!cl->receive(updates)) {
      throw std::runtime_error(std::string("Receive failure: ") + cl->error());
    }

    const auto state = cl->state();
    const auto myUnits = state->units[state->player_id];
    std::vector<tc::Client::Command> actions;
    nloop++;

    if (state->game_ended) {
      break;
    } else if (state->battle_just_ended) {
      if (state->battle_won) {
        battlesWon++;
      }
      battlesGame++;
      (*totalBattles)++;
      if (battlesGame >= 10) {
        actions.emplace_back(tc::BW::Command::Restart);
      }
      goto send;
    } else if (state->waiting_for_restart) {
      // Do nothing
      goto send;
    } else if (state->battle_frame_count % FLAGS_skip_frames != 0) {
      // Skip this frame
      goto send;
    }

    // Determine actions for each unit
    for (auto unit : myUnits) {
      auto utype = tc::BW::UnitType::_from_integral(unit.type);
      if (tc::BW::isBuilding(utype)) {
        if (utype == +tc::BW::UnitType::Terran_Barracks) {
          // Train a new marine unit
          actions.emplace_back(
              tc::BW::Command::CommandUnit,
              unit.id,
              tc::BW::UnitCommandType::Train,
              tc::BW::UnitType::Terran_Marine);
        }
      } else if (tc::BW::isWorker(utype)) {
        // Build barracks if ore count allows for it
        if (state->frame->resources[state->player_id].ore >= 150 &&
            state->frame_from_bwapi - builtBarracks > 240) {
          builtBarracks = state->frame_from_bwapi;
          auto ccenter = std::find_if(
              myUnits.begin(), myUnits.end(), [](tc::replayer::Unit u) {
                return u.type == tc::BW::UnitType::Terran_Command_Center;
              });
          if (ccenter != myUnits.end()) {
            // Already executing build order?
            if (!isExecutingCommand(unit, tc::BW::UnitCommandType::Build) &&
                !isExecutingCommand(
                    unit, tc::BW::UnitCommandType::Right_Click_Position)) {
              actions.emplace_back(
                  tc::BW::Command::CommandUnit,
                  unit.id,
                  tc::BW::UnitCommandType::Build,
                  -1,
                  ccenter->x,
                  ccenter->y + 8,
                  tc::BW::UnitType::Terran_Barracks);
            }
          }
        } else {
          // Gather resources
          if (!isExecutingCommand(unit, tc::BW::UnitCommandType::Gather) &&
              !isExecutingCommand(unit, tc::BW::UnitCommandType::Build) &&
              !isExecutingCommand(
                  unit, tc::BW::UnitCommandType::Right_Click_Position)) {
            auto mineralFields = getMineralFields(state);
            auto target = getClosest(
                unit.x, unit.y, mineralFields.begin(), mineralFields.end());
            actions.emplace_back(
                tc::BW::Command::CommandUnit,
                unit.id,
                tc::BW::UnitCommandType::Right_Click_Unit,
                target->id);
          }
        }
      } else {
        // Attack closest enemy
        auto enemyUnits = state->units[1 - state->player_id];
        auto target =
            getClosest(unit.x, unit.y, enemyUnits.begin(), enemyUnits.end());
        if (target != enemyUnits.end()) {
          actions.emplace_back(
              tc::BW::Command::CommandUnitProtected,
              unit.id,
              tc::BW::UnitCommandType::Attack_Unit,
              target->id);
        }
      }
    }

  // Send actions
  send:
    if (!cl->send(actions)) {
      throw std::runtime_error(std::string("Send failure: ") + cl->error());
    }
  }
}

// Program entry point
int main(int argc, char** argv) {
  torchcraft::init();
  gflags::ParseCommandLineFlags(&argc, &argv, true);

  int totalBattles = 0;
  while (totalBattles < 40) {
    std::cout << std::endl << "CTRL-C to stop" << std::endl << std::endl;

    try {
      auto cl = makeClient();
      playGame(std::move(cl), &totalBattles);
    } catch (std::exception& e) {
      std::cerr << e.what() << std::endl;
      return EXIT_FAILURE;
    }

    std::this_thread::sleep_for(std::chrono::milliseconds(500));
    std::cout << std::endl;
  }

  return EXIT_SUCCESS;
}
